api/control/v1alpha1/types.go

```diff
@@ -194,74 +194,54 @@ type Binding struct {
 // of a BindingPolicy's `what` and `where`: what objects to propagate and to where.
 // All objects present in this spec are propagated to all destinations present.
 type BindingSpec struct {
-	// `Workload` is a collection of namespaced-scoped objects and a collection of cluster-scoped objects to be propagated to destination clusters.
+	// `workload` is a collection of namespaced-scoped objects and a collection of cluster-scoped objects to be
+	// propagated to destination clusters.
 	Workload DownsyncObjectReferences `json:"workload,omitempty"`
 
 	// `destinations` is a list of cluster-identifiers that the objects should be propagated to.
 	Destinations []Destination `json:"destinations,omitempty"`
 }
 
-// DownsyncObjectReferences explicitly defines the objects to be down-synced.
-// The ClusterScope list defines the cluster-scope objects, NamespaceScope list
-// defines the namespace-scope objects and WorkloadGeneration represents the
-// generation of the objects in the ClusterScope and NamespaceScope lists.
-// Upon a change in any of workload objects that should be distributed
-// (e.g., spec, annotations or labels) the workload generation field should be incremented.
+// DownsyncObjectReferences defines the objects to be down-synced, grouping them by scope.
+// It specifies a set of object references, each associated with a ResourceVersion indicating
+// the specific version of the object to be downsynced. This effectively acts as a map from
```

Comment:
Currently the transport implementation does not check for the version when wrapping objects, but it should, later. I'll open an issue on the matter, but it should not block the release.

Reply:
It would be good if the comment could describe the design. What is the transport controller supposed to do with the ResourceVersion number found here? The current text here says that the ResourceVersion found here indicates which version to transport. But the transport controller can not request to read a specific version of an object. The apiservers store only one version of an object at a time.

The only thing I see that the transport controller _could_ do with the ResourceVersion here is compare it to the ResourceVersion already in the wrapped object to determine whether to wrap again.

I am not sure whether that is your intended design.

Reply:
If the object present in the WDS does not match the resource-version specified by the Binding, then the transport should (later) not transport the object. This would happen if the binding controller is not up to date with the state of the WDS, and should be resolved under eventual consistency. @nirrozenbaum

Reply:
You could take that view, but it seems unneccessarily fragile to me. It seems to be intended for a system where things proceed slowly, with both controllers fully reacting to each change of an object before another change happens. A more Kubernetes-native approach would be to not consider the ResourceVersion to be part of an object's identifier but rather a part of current state. In this approach, the transport controller would read the latest version (the only version available) of a referenced object whenever prodded to.

Reply:
It is true that with and without this constraint, eventual consistency takes care of the matter. I have removed:
 > ensuring that for each object, only a specific version is considered for downsyncing.
 
If you think the description should be different, please propose one.

Reply:
I tend to agree with @MikeSpreitzer perspectice. the previously used “workloadGeneration” was used as signaling mechanism, we now replaced it with resourceVersion per object, but the design remains the same. from transport point of view this is still a signaling mechanism. this was at least my view.
upon an event transport reads the latest version, not the version specified in the binding.

using this field per object instead of generation or revision that we maintain like we previously did is much more robust and works well even on pod crash/restarts.

Reply:
having said the above, IMO there’s no need to open issue for this, because this is the desired behavior.

Reply:
I have another question about what the transport controller should do if it finds that the current ResourceVersion of an object is not the ResourceVersion listed in the Binding. Should the transport controller: (a) revise the wrapped object to include only those whose ResourceVersion equals the one specified in the Binding, or (b) not update the wrapped object? If (a), will this cause the object with mismatched ResourceVersion to be deleted from the WEC? That would be the logical result, I think; and it would be bad. If (b) then this allows a series of badly timed changes to prevent the transport controller from ever transporting again.

Reply:
none of these options. 
as you previously mentioned- A more Kubernetes-native approach would be to not consider the ResourceVersion to be part of an object's identifier but rather a part of the object's variable state. In this approach, the transport controller would read the latest version (the only version available) of a referenced object.

so the resource version serves two purposes- (1) maintain what are the versions of objects that caused the creation/update of a binding, using resource version is robust to binding controller pod restarts and crashes; and (2) signaling mechanism between binding and transport controllers.

Reply:
from transport point of view we switched grom option (2) to option (1) from the discussion in #1671

Reply:
@nirrozenbaum : I think you are agreeing with my preferred semantics. My concerns here are that @vMaroon seems to be disagreeing and the current draft of this PR has line 207 reading more like what Maroon said than what you said: this line says that the ResourceVersion indicates the object version "to be downsynced".

Reply:
Also: the comments on BindingSpec and BindingSpec.Workload need to be accurate, and the comment on BindingSpec is the reader's entry point, should not be vague on this point.

Reply:
yes, I agree with your semantics. 
this is what transport code does :)
if the comments are not accurate we will fix to describe the semantics we just agreed on